tests/net/mod.rs

```diff
@@ -0,0 +1,782 @@
+                adversary.tamper(self, msg)
+            });
+            mem::replace(&mut self.adversary, adv);
+
+            // A missing adversary here could technically be a panic, but is impossible since we
+            // initialize with a `NullAdversary` upon construction.
+            try_some!(
+                opt_tamper_result.expect("No adversary defined (expected at least NullAdversary)")
+            )
+        } else {
+            // A correct node simply handles the message.
+            try_some!(self.dispatch_message(msg))
+        };
+
+        // All messages are expanded and added to the queue. We opt for copying them, so we can
+        // return unaltered step later on for inspection.
+        self.message_count = self.message_count.saturating_add(process_step(
+            &mut self.nodes,
+            receiver.clone(),
+            &step,
+            &mut self.messages,
+        ));
+
+        // Increase the crank count.
+        self.crank_count += 1;
+
+        Some(Ok((receiver, step)))
+    }
+
+    /// Convenience function for cranking.
+    ///
+    /// Shortcut for cranking the network, expecting both progress to be made as well as processing
+    /// to proceed.
+    pub fn crank_expect(&mut self) -> (D::NodeUid, Step<D>) {
+        self.crank()
+            .expect("crank: network queue empty")
+            .expect("crank: node failed to process step")
+    }
+}
+
+impl<D> VirtualNet<D>
+where
+    D: DistAlgorithm,
+    D::Message: Clone,
+    D::Input: Clone,
+    D::Output: Clone,
+{
+    /// Send input to all nodes.
+    ///
+    /// Equivalent to sending the same input to all nodes in order. Returns a vector of the
+    /// resulting `Step`s, which have had their messages queued already.
+    ///
+    /// If an error occurs, the first error will be returned and broadcasting aborted.
+    #[inline]
+    pub fn broadcast_input<'a>(
+        &'a mut self,
+        input: &'a D::Input,
+    ) -> Result<Vec<(D::NodeUid, Step<D>)>, D::Error> {
+        // Note: The tricky lifetime annotation basically says that the input value given must
+        //       live as long as the iterator returned lives (because it is cloned on every step,
+        //       with steps only evaluated each time `next()` is called. For the same reason the
+        //       network should not go away ealier either.
+
+        let steps: Vec<_> = self
+            .nodes
+            .values_mut()
+            .map(move |node| Ok((node.id().clone(), node.algorithm.input(input.clone())?)))
+            .collect::<Result<_, _>>()?;
+
+        // Process all messages from all steps in the queue.
+        steps.iter().for_each(|(id, step)| {
+            self.message_count = self.message_count.saturating_add(process_step(
+                &mut self.nodes,
+                id.clone(),
+                step,
+                &mut self.messages,
+            ));
+        });
+
+        Ok(steps)
```

Comment:
Would it be easier to call `send_input` in a loop?

Reply:
It would, but iterating over all node IDs borrows `self` by borrowing either `self.nodes` or `self.nodes()`. At that point, it's not possible to call `self.send_input` anymore.

The implementation here avoids having to copy the list of IDs and as a bonus sends all inputs "at once", that is steps are only processed once all nodes have sent theirs.